llvm/lib/Transforms/Utils/LoopIterCountAssumptions.cpp

```diff
@@ -158,10 +173,48 @@ void insertMinIterAssumption(ICmpInst &LoopCmpInstr, BasicBlock &LoopHeader,
                     << "MinIterValue                : " << *LoopAssertValue
                     << "\n");
 
-  Value *LHS = BunldedSCEV.IVOperandIndex == 0 ? LoopAssertValue
-                                               : LoopCmpInstr.getOperand(0);
-  Value *RHS = BunldedSCEV.IVOperandIndex == 0 ? LoopCmpInstr.getOperand(1)
-                                               : LoopAssertValue;
+  Value *Bound = BunldedSCEV.IVOperandIndex == 0 ? LoopCmpInstr.getOperand(1)
+                                                 : LoopCmpInstr.getOperand(0);
+
+  // Decouple the Bound and the ICmp Instruction by inserting 2 reversable
+  // Adds. If the Bound consists of a Load Instruction followed by a Constant
+  // offset, Correlated Value Propagation can deduce
+  // from the Assumption what the Branch Condition will evaluate to.           \
+  //       Conditional branch non-loop case \
+  // %i0 = add i32 0, 0                                                        \
+  // %0 = load i32, ptr p                                                      \
+  // %sub0 = add i32, %0, -1                                                   \
```

Comment:
Is this condition present in only one case?

Reply:
it occurs everywhere we have 

```
void function(unsigned &bound) {
  for (unsigned i=0; i < bound -1; i++) {
     //...
  }
}
```